docs/places.mdown

```diff
@@ -0,0 +1,107 @@
+
+## Autocomplete matching
+
+`history_api::search_frecent` returns all matches by origin, URL fragment, past autocomplete input, title, and tags, ranked by frecency and relevance.
+
+`history_api::match_url` returns the highest-ranked autocomplete match for an origin or URL fragment.
+
+`history_api::accept_result` records an input that completed to the chosen URL. This is stored and used in `search_frecent` to return matches by past autocomplete input.
+
+## Clearing history
+
+`history::delete_place_by_guid` deletes a URL and all its visits, and writes a Sync tombstone if the URL was synced. Bookmarked URLs can't be deleted until the bookmark is either deleted, or changed to point to a different URL.
+
+`history::delete_visits_between` deletes visits to a URL in a time range. If all visits to the URL fall within the range, the URL is also removed and tombstoned, except if it's bookmarked.
+
+`history::delete_place_visit_at_time` deletes a visit to a URL at a specific time. If it's the last remaining visit, the URL is also removed and tombstoned, except if it's bookmarked.
+
+`history::prune_destructively` prunes old history visits, without writing tombstones or Sync metadata. It's recommended to call this as a last resort—for example, if the device is low on disk space. Old visits _may_ be resurrected through Sync; however, since we only sync a subset of history, and expire old synced history after 60 days, this may cause data loss.
+
+`history::wipe_local` erases all history, without writing tombstones or Sync metadata. This means that visits from other devices can reappear over time, and a full sync is likely to resurrect all cleared history. `wipe_local` is meant for cases like clearing local data when signing out of Sync.
+
+Bookmarks
+=========
+
+The bookmarks API provides methods for creating, organizing, and deleting bookmark items: bookmarks, folders, and separators.
+
+## Creating bookmarks
+
+`bookmarks::insert_bookmark` inserts a bookmark item.
+
+`bookmarks::insert_tree` inserts a complete bookmark tree.
+
+## Getting bookmarks
+
+`bookmarks::get_bookmark` fetches info for a bookmark item.
+
+## Organizing bookmarks
+
+`bookmarks::fetch_tree` fetches info for a bookmark folder and all its descendants, recursively.
+
+`bookmarks::update_bookmark` updates or moves a bookmark item.
+
+## Deleting bookmarks
+
+`bookmarks::delete_bookmark` deletes a bookmark item. Folders are deleted recursively.
+
+Sync
+====
+
+Rust Places integrates syncing into the data model.
+
+Each syncable data type provides a store that describes how to marshal and unmarshal Sync records. A store implements methods for staging incoming records, merging, and recording telemetry. Each data type handles application differently. For example, history sync builds a plan that writes URLs and visits directly into the live Places database, while bookmarks are first written into a separate table that's then merged in to the live table.
```

Comment:
It's not clear from reading whether this divergence is deliberate or just an accident of history; what's the rationale for this divergence?

Reply:
My impression is, there are around a few ways we've done this. @linacambridge should correct me if I'm wrong (I probably am about bookmarks), I'm mostly trying to give the historical perspective.

1. Apply records as they come (most desktop sync engines). This has the downsides of failed syncs applying some of the state.
2. Do all queries up front, do a bunch of thinking in memory, build up a plan (either in memory, for logins, or in the db, for bookmarks) and apply the results at the end: This approach is fairly fast, relatively to reason about, and relatively easy to implement for engines like logins (I'd take this approach in the future for this sort of sync engine as well).
3. Do... Some of the queries up front, storing results in memory, and some as we go (Places history sync), but don't apply any changes until the end. This is easier to implement for history (due to it's structure), and (mostly) doesn't have the drawback of 1. It has some perf downsides compared to 2 though.

I think it's mostly an accident that history sync takes approach 3.

Reply:
Given unbounded time and resources, would we consider migrating history to use approach 2?

Reply:
Only for performance reasons, I think. It's fine as it is, but requires a lot more smaller queries (instead of a few batch queries). It does require more data be kept in memory at once, so I'm not 100% sure though.

It's also worth noting that logins and history are the same according to the description of 'builds a plan that writes URLs and visits directly into the live ... database', so there are multiple angles to look at this from. I don't know how much of this is worth calling out in the doc.